src/diffusers/models/unet_2d_blocks.py

```diff
@@ -1384,6 +1456,199 @@ def forward(
+
+                hidden_states = torch.utils.checkpoint.checkpoint(create_custom_forward(resnet), hidden_states, temb)
+            else:
+                hidden_states = resnet(hidden_states, temb)
+
+            output_states += (hidden_states,)
+
+        if self.downsamplers is not None:
+            for downsampler in self.downsamplers:
+                hidden_states = downsampler(hidden_states)
+
+        return hidden_states, output_states
+
+
+class KCrossAttnDownBlock2D(nn.Module):
+    def __init__(
+        self,
+        in_channels: int,
+        out_channels: int,
+        temb_channels: int,
+        cross_attention_dim: int,
+        dropout: float = 0.0,
+        num_layers: int = 4,
+        resnet_groups: Optional[int] = None,
+        resnet_group_size: int = 32,
+        add_downsample=True,
+        attn_num_head_channels: int = 64,
+        attn1_type: str = "cross",  # cross, self
+        attn2_type: Optional[str] = None,  # None, cross, self
+        resnet_eps: float = 1e-5,
+        resnet_act_fn: str = "gelu",
+        resnet_time_scale_shift: str = "scale_shift",
+    ):
+        super().__init__()
+        resnets = []
+        attentions = []
+
+        self.has_cross_attention = True
+        self.attn2_type = attn2_type
+
+        for i in range(num_layers):
+            in_channels = in_channels if i == 0 else out_channels
+            resnets.append(
+                ResnetBlock2D(
+                    in_channels=in_channels,
+                    out_channels=out_channels,
+                    dropout=dropout,
+                    temb_channels=temb_channels,
+                    groups=resnet_groups,
+                    eps=resnet_eps,
+                    non_linearity=resnet_act_fn,
+                    norm1_scale_shift=True,
+                    time_embedding_norm=resnet_time_scale_shift,
+                    dropout_after_conv=True,
+                    group_size=resnet_group_size,
+                    conv_shortcut_bias=False,
+                )
+            )
+            attentions.append(
+                BasicTransformerBlock(
+                    out_channels,
+                    out_channels // attn_num_head_channels,
+                    attn_num_head_channels,
+                    dropout=None,
+                    cross_attention_dim=cross_attention_dim,
+                    temb_channels=temb_channels,
+                    attention_bias=True,
+                    attn1_type=attn1_type,
+                    attn2_type=attn2_type,
+                    with_ff=False,
+                    use_conv_proj=True,
+                    cross_attention_norm=True,
+                    attention_dropout=dropout,
+                    group_size=resnet_group_size,
+                    norm_type="ada_group",
+                )
+            )
+
+        self.resnets = nn.ModuleList(resnets)
+        self.attentions = nn.ModuleList(attentions)
+
+        if add_downsample:
```

Comment:
@yiyixuxu note how the unet block here "absorbs" the complexity so that we don't have to add a `downsample` list to `Unet2DCondition`